### PR TITLE
Provide an abstraction over the lock file format

### PR DESCRIPTION
**Pull request by shs96c**
This starts laying the groundwork for us to replace the lock file format with something else without needing to rewrite everything from scratch.

### CODE REVIEW

**Comment by shs96c** [comment]
There is a required follow-up PR to handle the unpinned repo case, which will allow us to delete some of the code in `coursier.bzl` that is currently duplicated in the `v1_lock_file.bzl` file.

File: private/rules/v1_lock_file.bzl [hunk 1]
```diff
@@ -0,0 +1,166 @@
+def _remove_auth_from_url(url):
```
Comment (jin) [inline private/rules/v1_lock_file.bzl:1]: nit: license header

File: private/rules/v1_lock_file.bzl [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    protocol, url_parts = _split_url(url)
+    host = url_parts[0]
+    if "@" not in host:
+        return url
+    last_index = host.rfind("@", 0, None)
+    userless_host = host[last_index + 1:]
+    new_url = "{}://{}".format(protocol, "/".join([userless_host] + url_parts[1:]))
+    return new_url
+
+def _is_valid_lock_file(lock_file_contents):
```
Comment (jin) [inline private/rules/v1_lock_file.bzl:14]: Should we also apply a recursive spec check on the expected keys in the dependency_tree object?

**Review by jin — APPROVED** [review]
thanks! lgtm

File: private/rules/v1_lock_file.bzl (hunk 1 above)
Comment (shs96c) [reply private/rules/v1_lock_file.bzl:14]: Added

File: private/rules/v1_lock_file.bzl (hunk 2 above)
Comment (shs96c) [reply private/rules/v1_lock_file.bzl:27]: Agreed. This does need more checks in it :) Done!

**Review by jin — APPROVED** [review]

**Pull request merged**